### PR TITLE
Step by step

### PR DESCRIPTION
**Pull request by Ho-Ro**
As proposed here are the step by step commits, please check

### CODE REVIEW

**Comment by Ho-Ro** [comment]
@ribalda I think the tool is now in a complete state - please check!

**Review by ribalda — COMMENTED** [review]
I'd rather not merge the "Turn into stand-alone tool" patch to avoid all the licensing issues. I am not familiar enough with the matter.

**Review by ribalda — DISMISSED** [review]

File: fx2eeprom.c [hunk 1]
```diff
@@ -44,31 +44,37 @@
...
+	fprintf(stdout,"%s w/r VID PID [ size [ee_addr] ]\n",prog);
 	return;
 }
 
 int main(int argc, char *argv[]){
 	struct libusb_device_handle *dev;
 	int ret;
-	unsigned char *buffer;
-	int pid, vid, length;
+	unsigned char buffer[CHUNKSIZE];
```
Comment (ribalda) [inline fx2eeprom.c:64]: I prefer to keep the memory allocation, it simplifies the code.

File: fx2eeprom.c [hunk 2]
```diff
@@ -123,10 +126,13 @@ int main(int argc, char *argv[]){
 	}
 	else {
 		ret=fread(buffer,1,length,stdin);
-		if (length!=ret){
-			fprintf(stderr,"Wrong size from stdin, expected %d read %d\n",length,ret);
-			perror("fread");
-			return -1;
+		if (ret < length){ /* too few bytes? */
```
Comment (ribalda) [inline fx2eeprom.c]: What if fread returns an error?. I would leave the check that we currently have. If the user provided a file of a smaller size it should be an error.

File: fx2eeprom.c [hunk 3]
```diff
@@ -60,19 +60,23 @@ int main(int argc, char *argv[]){
 	int address=0;
 	int mode;
 
-	if (argc!=5){
+	if (argc<4){
 		use(argv[0]);
 		return -1;
 	}
 
+	if (argc==5)
```
Comment (ribalda) [inline fx2eeprom.c]: Why do you want to provide a default memory size? Doesnt it depend on the chip model?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+`vend_ax.hex` is property of Cypress.
+
+COMPILE
+-------
+- Install libusb-1.0-0-dev on your system, e.g. `apt install libusb-1.0-0-dev` for debian.
+- Run `make`
+
+PREPARATION
+-----------
+- Install cycfx2prog, e.g. `apt install cycfx2prog`
```
Comment (ribalda) [inline README.md]: isnt fxload more common?

File: vend_ax.c [hunk 5]
```diff
@@ -0,0 +1,318 @@
+/***************************************************
```
Comment (ribalda) [inline vend_ax.c:1]: I'd rather avoid this patch completely. Not familiar enough with the licensing implications.

Comment (Ho-Ro) [reply vend_ax.c:1]: This comes from the cypress program cyusb suite for linux (e.g. [my fork](https://github.com/Ho-Ro/cyusb_linux), esp. from [src/download_fx2.cpp](https://github.com/Ho-Ro/cyusb_linux/blob/main/src/download_fx2.cpp)), that is under LGPL.
Also you provide the same code - but in hex format as `vend_ax.hex`.

In short: A combination of GPL code and LGPL code must be licensed under GPL.
LGPL code has the same license as GPL with the addition that you can build a library out of it and share it _WITHOUT_ the need to provide the source code, this takes away the viral GPL effect. This additional freedom (as seen by some developers) is rewoked as soon GPL code is added.

File: README.md (hunk 4 above)
Comment (Ho-Ro) [reply README.md]: It uses the deprecated `usbfs` file system, while cycfx2prog uses VID PID.

File: fx2eeprom.c (hunk 3 above)
Comment (Ho-Ro) [reply fx2eeprom.c]: The memory size is most important for read operation, if I e.g. want to get the 1st 8 bytes from a C0 section. But also writing can be sized, e.g. `fx2eeprom w $VID $PID 8192 </dev/zero` to erase the EEPROM. This is also the reason I invented the EEPROM memory address as 5th optional parameter.

File: fx2eeprom.c (hunk 2 above)
Comment (Ho-Ro) [reply fx2eeprom.c]: This requires that you know the file size, in Makefiles it needs a previous step with `SIZE=$(wc -c ${INFILE})`.

File: fx2eeprom.c (hunk 1 above)
Comment (Ho-Ro) [reply fx2eeprom.c:64]: Are you able to work with files > 4K bytes? I got failures when I did so, while reading/writing chunkwise worked also with 16K byte (as well as 64K where I got 4x mirrored data).

**Comment by Ho-Ro** [comment]
Ok, understood, it is your project.
I changed it to my needs but this is on a different page.
Thank you for the code base.

Feel free to close this PR.

Martin

**Comment by Ho-Ro** [comment]
> I'd rather avoid this patch completely. Not familiar enough with the licensing implications.

Just in case anyone is watching this conversation - the patch is fully covered by the GNU licences, which allow you to use LGPLed code and incorporate it into your own or a derivative work, it's called *"copy code "*:

> When we say “copy code,” we mean just that: you're taking a section of code from one source, with or without modification, and inserting it into your own program, thus forming a work based on the first section of code.

This is explained in detail in this document:
[How are the various GNU licenses compatible with each other?](https://www.gnu.org/licenses/gpl-faq.html#AllCompatibility)

image

However, I am continuing to develop the program on my fork https://github.com/Ho-Ro/fx2eeprom.
I am therefore withdrawing my PR here.

**Pull request closed without merging**